src/Dependencies/PooledObjects/PooledDelegates.cs

```diff
@@ -0,0 +1,404 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+
+namespace Microsoft.CodeAnalysis.PooledObjects
+{
+    /// <summary>
+    /// Provides pooled delegate instances to help avoid closure allocations for delegates that require a state argument
+    /// with APIs that do not provide appropriate overloads with state arguments.
+    /// </summary>
+    internal static class PooledDelegates
+    {
+        private static class DefaultDelegatePool<T>
+            where T : class, new()
+        {
+            public static readonly ObjectPool<T> Instance = new ObjectPool<T>(() => new T(), 20);
+        }
+
+        private static Releaser GetPooledDelegate<TPooled, TArg, TUnboundDelegate, TBoundDelegate>(TUnboundDelegate unboundDelegate, TArg argument, out TBoundDelegate boundDelegate)
+            where TPooled : AbstractDelegateWithBoundArgument<TArg, TUnboundDelegate, TBoundDelegate>, new()
+            where TUnboundDelegate : Delegate
+            where TBoundDelegate : Delegate
+        {
+            var obj = DefaultDelegatePool<TPooled>.Instance.Allocate();
+            obj.Argument = argument;
+            obj.UnboundDelegate = unboundDelegate;
+            boundDelegate = obj.BoundDelegate;
+
+            return new Releaser(
+                obj,
+                obj =>
+                {
+                    var pooledObject = (TPooled)obj;
+                    pooledObject.Clear();
+                    DefaultDelegatePool<TPooled>.Instance.Free(pooledObject);
+                });
+        }
+
+        /// <summary>
+        /// Gets an <see cref="Action"/> delegate, which calls <paramref name="unboundAction"/> with the specified
```

Comment:
>Gets an <see cref="Action"/> delegate, which calls <paramref name="unboundAction"/> with the specified [](start = 12, length = 102)

It is not clear why would one want to use this API. What is the advantage vs. "normal" way of creating a delegate? How is it achieved? If something is getting cashed somewhere. What? For how long? Etc.

Reply:
> It is not clear why would one want to use this API. What is the advantage vs. "normal" way of creating a delegate?

This is covered by the documentation for the type itself:

_Provides pooled delegate instances to help avoid closure allocations for delegates that require a state argument with APIs that do not provide appropriate overloads with state arguments._

The type does not make any guarantees with respect to the internal cache behavior as it relates to any particular call. The API only guarantees that the resulting delegate will behave the same way a by-value state argument behaves, from the point at which the delegate is obtained until the point at which the releaser is disposed.

A user should consider using this API if/when frequent allocation of delegates for callbacks at some location outweighs the maintenance risk introduced by using a pooled mutable object (specifically the risk of use-after-free).

> If something is getting cashed somewhere. What? For how long? Etc.

The type does not cache any user-provided arguments, and therefore does not impact the GC lifetimes of any of those objects. The internal cache semantics are inherited from `ObjectPool<T>`.